### PR TITLE
faiss: 1.7.4 -> 1.8.0

### PR DESCRIPTION
**Pull request by SomeoneSerge**
(cherry picked from commit 39dc5d0d768ec5cb51636b449f08fefd96c70745)

Trying to recover @bcdrwin's https://github.com/NixOS/nixpkgs/pull/313192

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by afh** [comment]
Thanks for picking this up, @SomeoneSerge, much appreciated 🙏

**Comment by afh** [comment]
Result of `nixpkgs-review pr 330191` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>sqlite-vss</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>faiss</li>
    <li>faiss.demos</li>
  </ul>
</details>

**Comment by afh** [comment]
Regarding the build failures:
1. `sqlite-vss` fails with `/tmp/nix-build-sqlite-vss-0.1.2.drv-0/source/src/sqlite-vss.cpp:1245:25: error: no type named 'FaissException' in namespace 'faiss'` so it might not be compatible with faiss 1.8.0. According to [upstream](https://github.com/asg017/sqlite-vss):
> sqlite-vss is not in active development. Instead, my effort is now going towards [sqlite-vec](https://github.com/asg017/sqlite-vec), which is a similar vector search SQLite extension, but should be much easier to install and use than sqlite-vss. See [this blog post](https://alexgarcia.xyz/blog/2024/building-new-vector-search-sqlite/index.html) for more info.
2. whisper-ctranslate2 and whisper-ctranslate2.dist fail due to strictly requiring gpytorch 0.11, yet only version 0.12 is available in nixpkgs, this is addressed in #328732

File: pkgs/development/libraries/science/math/faiss/default.nix [hunk 1]
```diff
@@ -51,18 +51,9 @@ stdenv.mkDerivation {
     owner = "facebookresearch";
     repo = pname;
     rev = "v${version}";
-    hash = "sha256-WSce9X6sLZmGM5F0ZkK54VqpIy8u1VB0e9/l78co29M=";
+    hash = "sha256-nS8nhkNGGb2oAJKfr/MIAZjAwMxBGbNd16/CkEtv67I=";
   };
 
-  patches = [
-    (fetchpatch2 {
```
Comment (JohnRTitor) [inline pkgs/development/libraries/science/math/faiss/default.nix:58]: Remove `fetchpatch2` from top level as well.

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 330191 --extra-nixpkgs-config '{
  cudaSupport = true;
  cudaCapabilities = [
    "7.0"
    "8.0"
    "8.6"
  ];
  cudaEnableForwardCompat = false;
  allowUnfreePredicate =
    p:
    builtins.all (
      license:
      license.free
      || builtins.elem license.shortName [
        "CUDA EULA"
        "cuDNN EULA"
        "cuTENSOR EULA"
        "NVidia OptiX EULA"
      ]
    ) (if builtins.isList p.meta.license then p.meta.license else [ p.meta.license ]);
  permittedInsecurePackages = [
    "electron-24.8.6"
    "zotero-6.0.27"
  ];
}'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>sqlite-vss</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>faiss (faissWithCuda ,python312Packages.faiss)</li>
    <li>faiss.demos (faissWithCuda.demos ,python312Packages.faiss.demos)</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.demos</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
  </ul>
</details>

**Comment by SomeoneSerge** [comment]
The whole expression needs to be refactored and maybe the pip-install bit moved into a separate derivation, but I'll do this refactoring some time later. Can't publish the logs for whisper and stuff because Nix only kept logs for the succeeding derivations (very smart)

**Comment by SomeoneSerge** [comment]
> whisper-ctranslate2

Ok, it's not me, it's botorch. IIRC there's an open PR addressing that?

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 330191 --extra-nixpkgs-config '{
  cudaSupport = true;
  cudaCapabilities = [
    "7.0"
    "8.0"
    "8.6"
  ];
  cudaEnableForwardCompat = false;
  allowUnfreePredicate =
    p:
    builtins.all (
      license:
      license.free
      || builtins.elem license.shortName [
        "CUDA EULA"
        "cuDNN EULA"
        "cuTENSOR EULA"
        "NVidia OptiX EULA"
      ]
    ) (if builtins.isList p.meta.license then p.meta.license else [ p.meta.license ]);
  permittedInsecurePackages = [
    "electron-24.8.6"
    "zotero-6.0.27"
  ];
}'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>sqlite-vss</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>faiss (faissWithCuda ,python312Packages.faiss)</li>
    <li>faiss.demos (faissWithCuda.demos ,python312Packages.faiss.demos)</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.demos</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
  </ul>
</details>

**Comment by afh** [comment]
Rebasing this commit onto the latest master should at least get rid of the whisper-ctranslate2 failures now that #328732 is merged.

**Comment by SomeoneSerge** [comment]
> Rebasing this commit onto the latest master should at least get rid of the whisper-ctranslate2 failures now that https://github.com/NixOS/nixpkgs/pull/328732 is merged.

nixpkgs-review does the rebasing

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 330191 --extra-nixpkgs-config '{
  cudaSupport = true;
  cudaCapabilities = [
    "7.0"
    "8.0"
    "8.6"
  ];
  cudaEnableForwardCompat = false;
  allowUnfreePredicate =
    p:
    builtins.all (
      license:
      license.free
      || builtins.elem license.shortName [
        "CUDA EULA"
        "cuDNN EULA"
        "cuTENSOR EULA"
        "NVidia OptiX EULA"
      ]
    ) (if builtins.isList p.meta.license then p.meta.license else [ p.meta.license ]);
  permittedInsecurePackages = [
    "electron-24.8.6"
    "zotero-6.0.27"
  ];
}'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>sqlite-vss</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>faiss (faissWithCuda ,python312Packages.faiss)</li>
    <li>faiss.demos (faissWithCuda.demos ,python312Packages.faiss.demos)</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.demos</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
  </ul>
</details>

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 330191 --extra-nixpkgs-config '{
  cudaSupport = true;
  cudaCapabilities = [
    "7.0"
    "8.0"
    "8.6"
  ];
  cudaEnableForwardCompat = false;
  allowUnfreePredicate =
    p:
    builtins.all (
      license:
      license.free
      || builtins.elem license.shortName [
        "CUDA EULA"
        "cuDNN EULA"
        "cuTENSOR EULA"
        "NVidia OptiX EULA"
      ]
    ) (if builtins.isList p.meta.license then p.meta.license else [ p.meta.license ]);
  permittedInsecurePackages = [
    "electron-24.8.6"
    "zotero-6.0.27"
  ];
}'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.txtai</li>
    <li>python312Packages.txtai.dist</li>
    <li>sqlite-vss</li>
  </ul>
</details>
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>faiss (faissWithCuda ,python312Packages.faiss)</li>
    <li>faiss.demos (faissWithCuda.demos ,python312Packages.faiss.demos)</li>
    <li>nixpkgs-manual</li>
    <li>python311Packages.autofaiss</li>
    <li>python311Packages.autofaiss.dist</li>
    <li>python311Packages.faiss</li>
    <li>python311Packages.faiss.demos</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pytorch-metric-learning</li>
    <li>python311Packages.pytorch-metric-learning.dist</li>
    <li>python311Packages.txtai</li>
    <li>python311Packages.txtai.dist</li>
    <li>python312Packages.autofaiss</li>
    <li>python312Packages.autofaiss.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pytorch-metric-learning</li>
    <li>python312Packages.pytorch-metric-learning.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by SomeoneSerge** [comment]
https://github.com/NixOS/nixpkgs/pull/330191/checks?check_run_id=28002977367 refers to the (now missing) passthru tests, I guess keep them broken until we've figured out why they don't link

**Pull request merged**